File: HW9/Driver.c
/**
 *Name: ChakMan Sio, Mengyu Yang
 *Class: CSE12
 *Account: cs12xef, cs12xfh
 *Assignment: HW9
 *File Name: Driver.c
 *Description: this fiel handles the input to pup up different prompts and
 *			       call corresponding methods to run the program. 
 */

#include <iostream>
#include <cstdio>
#include <string>
#include <getopt.h>
#include "Driver.h"
#include "SymTab.h"
#include <fstream>
using namespace std;

#ifdef NULL
#undef NULL
#define NULL 0
#endif


ostream & operator << (ostream & stream, const UCSDStudent & stu) {
        return stream << "name:  " << stu.name
                << " with studentnum:  " << stu.studentnum;
}

/*
 *Name: main
 *Description: the main method to run the program, uses the switch case to 
 *			handle differnt input and then calling correspond methods
 *
 */
int main (int argc, char * const * argv) {
        char buffer[BUFSIZ];
        char command;
        long number;
        char option;
        
				/*initialize the input stream*/
		    istream * is;
		    is = &cin;
				/*initialize the output stream*/
		    ostream * os;
		    os = &cout;

        SymTab<UCSDStudent>::Set_Debug_Off ();

        while ((option = getopt (argc, argv, "x")) != EOF) {

        switch (option) {
                case 'x': SymTab<UCSDStudent>::Set_Debug_On ();
                        break;
                }
        }
        
				SymTab<UCSDStudent> ST("Driver.datafile");
        ST.Write (cout << "Initial Symbol Table:\n" );

       	while (cin) {
								
								if (!(*is)){
								    // delete the is
                    delete is;
										delete os;
										// is store the address of cin
                    is = &cin;
										// os store the address of cout
                    os = &cout;
								}
                command = NULL;         // reset command each time in loop
                *os << "Please enter a command ((f)ile, (i)nsert, "
                        << "(l)ookup, (r)emove, (w)rite):  ";
                *is >> command;

                switch (command) {
                        
                case 'i': {
                        *os << "Please enter UCSD student name to insert:  ";
                        *is >> buffer;  // formatted input

                        *os << "Please enter UCSD student number:  ";
                        *is >> number;

                        UCSDStudent stu (buffer, number);

                        // create student and place in symbol table
                        ST.Insert (stu);
                        break;
                }
								case 'l': {
                        unsigned long found;    // whether found or not

                        *os << "Please enter UCSD student name to lookup:  ";
                        *is >> buffer;  // formatted input

                        UCSDStudent stu (buffer, 0);
                        found = ST.Lookup (stu);
                        
                        if (found)
                                cout << "Student found!!!\n" << stu << "\n";
                        else
                                cout << "student " << buffer << " not there!\n";
                        break;
                }
                case 'r': { 
                        unsigned long removed;

                        *os << "Please enter UCSD student name to remove:  ";
                        *is >> buffer;  // formatted input

                        UCSDStudent stu (buffer, 0);
                        removed = ST.Remove(stu);

                        if (removed)
                                cout << "Student removed!!!\n" << stu << "\n";
                        else
                                cout << "student " << buffer << " not there!\n";
                        break;
                }
								/*the file case for this program*/
							  case 'f':{
												/*output prompt*/
                        *os <<"Please enter file name for commands";
                        *is >> buffer;  // formatted input
												/*transet to another file*/
												if (is !=&cin){
                            delete is;
														/*assign a new ifstream*/
												    is = new ifstream(buffer);
														break;
												}
                         
												/*new output stream to store buffer*/
												is = new ifstream(buffer);
												/*new input stream to store null*/
												os = new ofstream("dev/null");
												break;

								}

                case 'w':{
                        ST.Write (cout << "The Symbol Table contains:\n");
								}
                }
				}
		
        ST.Write (cout << "\nFinal Symbol Table:\n");
        /*calculate the efficiency of the program afterEOF*/      
					if (ST.GetOperation() != 0) {
				      /*get cost*/
				      cout << "\nCost of operations: ";
			        cout << ST.GetCost();
			        cout << " tree accesses";
              /*get opration*/
			        cout << "\nNumber of operations: ";
			        cout << ST.GetOperation();
				      /*calculate the average*/
			        cout << "\nAverage cost: ";
		        	cout << (((float)(ST.GetCost()))/(ST.GetOperation()));
			        cout << " tree accesses/operation\n";
		      }
			    /*no info to return*/
		      else
			        cout << "\nNo cost information available.\n";
}

File: HW7/Driver.java
/**
 * Name: ChakMan Sio 
 * Class: CSE 12 
 * Account cs12xef
 * Quarter: Fall 2016 
 * Assignment: HW7
 * File name: Driver.java
 * Class Description: this class allow to accsss the UCSDStudent, I can 
 *                    use those functions to implement my Tree.java
 */


import java.io.*;

class UCSDStudent extends Base {

        private String name;
        private long studentnum;

        public long value;              /* value of interest */

        public UCSDStudent (String nm, long val) {
                name = new String (nm);
                value = val;
        }
        public UCSDStudent (UCSDStudent  variable) {
                name = new String (variable.name);
                value = variable.value;
        }

        public String getName () {
                return name;
        }
        public boolean equals (Object object) {
		        if (this == object)
			          return true;

		        if (!(object instanceof UCSDStudent ))
			          return false;
		
		        UCSDStudent  otherVar = (UCSDStudent) object;
		
		        return name.equals (otherVar.getName ());
        }
        public boolean isLessThan (Base base) {
                return (name.compareTo (base.getName ()) < 0) ? true : false;
        }
        public String toString () {
                return "name: "+name+ " " + "Studentnum: " + value;
        }

        public UCSDStudent  assignValue (long val) {
                
                UCSDStudent  retval;        // return value

                // give variable its value
                value = val;
                retval = new UCSDStudent  (this);

                return retval;
        }

}

public class Driver {
        private static final short NULL = 0;

        public static void main (String [] args) {
        
                /* initialize debug states */
                Tree.debugOff();

                /* check command line options */
                for (int index = 0; index < args.length; ++index) {
                        if (args[index].equals("-x"))
                                Tree.debugOn();
                }

                /* The real start of the code */
                SymTab<UCSDStudent> symtab = 
                        new SymTab<UCSDStudent>("UCSDStudentTree");
                String buffer = null;
                char command;
                long number = 0;

                System.out.println ("Initial Symbol Table:\n" + symtab);

                while (true) {
                        command = NULL; // reset command each time in loop
                        System.out.print ("Please enter a command:  " + 
                                "((a)llocate, (i)nsert, (l)ookup, " +
                                "(r)emove, (w)rite):  ");

                        try {
                        command = MyLib.getchar ();
                        MyLib.clrbuf (command); // get rid of return

                        switch (command) {

                        case 'a':
                                System.out.print
                                        ("Please enter name of new Tree to " +
                                         "allocate:  ");
                                
                                buffer = MyLib.getline ();// formatted input
                                symtab = new SymTab<UCSDStudent>(buffer);
                                break;

                        case 'i':
                                System.out.print
                                ("Please enter UCSD student name to insert:  ");

                                buffer = MyLib.getline ();// formatted input

                                System.out.print
                                        ("Please enter UCSD student number:  ");

                                number = MyLib.decin ();
                                MyLib.clrbuf (command); // get rid of return

                                // create student and place in symbol table
                                symtab.insert(new UCSDStudent (buffer, number));
                                break;

                        case 'l': { 
                                UCSDStudent found;      // whether found or not

                                System.out.print
                                ("Please enter UCSD student name to lookup:  ");
                                buffer = MyLib.getline ();// formatted input

                                UCSDStudent stu = new UCSDStudent (buffer, 0);
                                found = symtab.lookup (stu);
                                
                                if (found != null) {
                                        System.out.println("Student found!!!");
                                        System.out.println(found);
                                }
                                else
                                        System.out.println ("student " + buffer
                                                + " not there!!!");
                                }
                                break;
                        
                        case 'r': { 
                                UCSDStudent removed; // data to be removed

                                System.out.print
                                ("Please enter UCSD student name to remove:  ");

                                buffer = MyLib.getline ();

                                UCSDStudent stu = new UCSDStudent (buffer, 0);

                                removed = symtab.remove(stu);

                                if (removed != null) {
                                        System.out.println("Student removed!"); 
                                        System.out.println(removed);
                                }
                                else
                                        System.out.println ("student " + buffer
                                                + " not there!!!");
                                }
                                break;

                        case 'w':
                                System.out.println("The Symbol Table " +
                                "contains:\n" + symtab);
                        }
                        }
                        catch (EOFException eof) {
                                break;
                        }
                }
                System.out.println("\nFinal Symbol Table:\n" + symtab);
        }
}
